plugins/inputs/t128_graphql/query_builder.go

```diff
@@ -1,20 +1,107 @@
 package t128_graphql
 
-import "strings"
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/Jeffail/gabs"
+)
 
 func buildQuery(entryPoint string, fields map[string]string, tags map[string]string) string {
-	var replacer = strings.NewReplacer("[", "(", "]", "\")", "/", "{", ":", ":\"")
-	query := "query {" + replacer.Replace(entryPoint) + "{"
+	query := "query "
+
+	var buf bytes.Buffer
+	w := io.Writer(&buf)
+	jsonObj := buildQueryObject(entryPoint, fields, tags)
+	traverseQueryObjPreOrder(jsonObj, w)
+
+	query += buf.String()
+	return query
+}
 
-	for _, element := range fields {
-		query += "\n" + element
+func traverseQueryObjPreOrder(jsonObj *gabs.Container, w io.Writer) {
+	jsonChildren, err := jsonObj.ChildrenMap()
+	if err != nil {
+		fmt.Println("error")
+		return
 	}
-	query = strings.TrimSpace(query)
-	for _, element := range tags {
-		query += "\n" + element
+
+	//store and sort the keys
+	keys := make([]string, len(jsonChildren))
+	i := 0
+	for k := range jsonChildren {
+		keys[i] = k
+		i++
 	}
+	sort.Strings(keys)
 
-	query += "\n" + strings.Repeat("}", strings.Count(query, "{"))
+	children := 0
+	for i, key := range keys {
+		if strings.HasPrefix(key, "$") {
+			w.Write([]byte("(" + key[1:] + ":\"" + jsonChildren[key].Data().(string) + "\"){"))
+			continue
+		}
+		children++
+		_, ok := jsonChildren[key].Data().(string)
+		if !ok {
+			if i == 0 {
+				w.Write([]byte("{"))
+			}
+			w.Write([]byte("\n" + key))
+			traverseQueryObjPreOrder(jsonChildren[key], w)
+			continue
+		}
+		if i == 0 {
+			w.Write([]byte("{"))
+		}
+		w.Write([]byte("\n" + key))
+	}
 
-	return query
+	if children > 0 {
+		w.Write([]byte("}"))
+	}
+}
+
+func buildQueryObject(entryPoint string, fields map[string]string, tags map[string]string) *gabs.Container {
+	jsonObj := gabs.New()
+
+	//add items in entry point to obj
+	path := ""
+	pathElements := strings.Split(entryPoint, "/")
+	for _, element := range pathElements {
```

Comment:
I'm assuming this was easier than what we discussed where we generate multiple gab paths. It is a little tough to read. I'd suggest either breaking it up into smaller functions or adding more intermediate variables to make it clear what you are doing. Either way, I'd suggest breaking this whole chunk into a seperate function so that you can do some unit testing on this complex part easier.